File: android/app/src/main/java/com/bwqr/mavinote/models/Account.kt
package com.bwqr.mavinote.models

import com.bwqr.mavinote.reax.Deserialize
import com.novi.serde.DeserializationError
import com.novi.serde.Deserializer

data class Account(
    val id: Int,
    val name: String,
    val kind: AccountKind,
) {
    companion object : Deserialize<Account> {
        override fun deserialize(deserializer: Deserializer): Account {
            deserializer.increase_container_depth()

            val account = Account(
                deserializer.deserialize_i32(),
                deserializer.deserialize_str(),
                AccountKind.deserialize(deserializer),
            )

            deserializer.decrease_container_depth()

            return account
        }
    }
}

enum class AccountKind {
    Mavinote,
    Local;

    companion object {
        fun deserialize(deserializer: Deserializer): AccountKind {
            val index = deserializer.deserialize_variant_index()

            return when (index) {
                0 -> Mavinote
                1 -> Local
                else -> throw DeserializationError("Unknown variant index for AccountKind: $index")
            }
        }
    }
}

data class Mavinote(val email: String, val token: String) {
    companion object : Deserialize<Mavinote> {
        override fun deserialize(deserializer: Deserializer): Mavinote {
            deserializer.increase_container_depth()

            val account = Mavinote(
                deserializer.deserialize_str(),
                deserializer.deserialize_str(),
            )

            deserializer.decrease_container_depth()

            return account
        }
    }
}

File: ios/Mavinote/Models/Account.swift
import Serde

struct Account : Identifiable {
    let id: Int32
    let name: String
    let kind: AccountKind

    static func deserialize(_ deserializer: Deserializer) throws -> Account {
        try deserializer.increase_container_depth()

        let account = Account(
            id: try deserializer.deserialize_i32(),
            name: try deserializer.deserialize_str(),
            kind: try AccountKind.deserialize(deserializer)
        )

        try deserializer.decrease_container_depth()

        return account
    }

}

enum AccountKind : String, CaseIterable {
    case Mavinote
    case Local

    static func deserialize(_ deserializer: Deserializer) throws -> AccountKind {
        let index = try deserializer.deserialize_variant_index()

        switch index {
        case 0: return .Mavinote
        case 1: return .Local
        default: throw DeserializationError.invalidInput(issue: "Unknown variant index for AccountKind")
        }
    }
}

struct Mavinote {
    let email: String
    let token: String

    static func deserialize(_ deserializer: Deserializer) throws -> Mavinote {
        try deserializer.increase_container_depth()

        let mavinote = Mavinote(
            email: try deserializer.deserialize_str(),
            token: try deserializer.deserialize_str()
        )

        try deserializer.decrease_container_depth()

        return mavinote

    }
}
